isobus/src/isobus_heartbeat.cpp

```diff
@@ -0,0 +1,287 @@
+//================================================================================================
+/// @file isobus_heartbeat.cpp
+///
+/// @brief Implements an interface for sending and receiving ISOBUS heartbeats.
+/// The heartbeat message is used to determine the integrity of the communication of messages and
+/// parameters being transmitted by a control function. There may be multiple instances of the
+/// heartbeat message on the network, and CFs are required transmit the message on request.
+/// As long as the heartbeat message is transmitted at the regular
+/// time interval and the sequence number increases through the valid range, then the
+/// heartbeat message indicates that the data source CF is operational and provides
+/// correct data in all its messages
+///
+/// @author Adrian Del Grosso
+///
+/// @copyright 2024 The Open-Agriculture Developers
+//================================================================================================
+#include "isobus/isobus/isobus_heartbeat.hpp"
+
+#include "isobus/isobus/can_general_parameter_group_numbers.hpp"
+#include "isobus/isobus/can_network_manager.hpp"
+#include "isobus/isobus/can_parameter_group_number_request_protocol.hpp"
+#include "isobus/isobus/can_stack_logger.hpp"
+#include "isobus/utility/system_timing.hpp"
+
+namespace isobus
+{
+	void HeartbeatInterface::initialize()
+	{
+		if (enabled && !initialized)
+		{
+			auto internalControlFunctions = CANNetworkManager::CANNetwork.get_internal_control_functions();
+
+			for (auto &cf : internalControlFunctions)
+			{
+				auto pgnRequestProtocol = cf->get_pgn_request_protocol().lock();
+
+				if (nullptr != pgnRequestProtocol)
+				{
+					pgnRequestProtocol->register_request_for_repetition_rate_callback(static_cast<std::uint32_t>(CANLibParameterGroupNumber::HeartbeatMessage), process_request_for_heartbeat, this);
+				}
+			}
+			CANNetworkManager::CANNetwork.add_any_control_function_parameter_group_number_callback(static_cast<std::uint32_t>(CANLibParameterGroupNumber::HeartbeatMessage), process_rx_message, this);
+			initialized = true;
+		}
+	}
+
+	bool HeartbeatInterface::get_is_initialized() const
+	{
+		return initialized;
+	}
+
+	void HeartbeatInterface::set_enabled(bool enable)
+	{
+		if ((!enable) && (enable != enabled))
+		{
+			LOG_DEBUG("[HB]: Disabling ISOBUS heartbeat interface.");
+			disable();
+		}
+		enabled = enable;
+	}
```

Comment:
resolving circular-dependency: remove call to `disable()` here, and instead use an early return of some kind in the `update()` and `process_request_for_heartbeat()`. Although not ideal, it simplifies the states disabling/enabling the interface and helps us remove the circular dependency
```c++
if (!enabled) {
    return;
}
```

Reply:
I prefer adding to the top of the enclosing `if` checks personally, which accomplishes the same thing basically. Added where needed.